File: src/main/java/com/javatechie/webflux/dao/CustomerDao.java
package com.javatechie.webflux.dao;

import com.javatechie.webflux.dto.Customer;
import org.springframework.stereotype.Component;
import reactor.core.publisher.Flux;

import java.time.Duration;
import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

@Component
public class CustomerDao {


    private static void sleepExecution(int i){
        try {
            Thread.sleep(1000);
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
    }

    public List<Customer> getCustomers()  {
        return IntStream.rangeClosed(1, 10)
                .peek(CustomerDao::sleepExecution)
                .peek(i -> System.out.println("processing count : " + i))
                .mapToObj(i -> new Customer(i, "customer" + i))
                .collect(Collectors.toList());
    }


    public Flux<Customer> getCustomersStream()  {
        return Flux.range(1,10)
                .delayElements(Duration.ofSeconds(1))
                .doOnNext(i -> System.out.println("processing count in stream flow : " + i))
                .map(i -> new Customer(i, "customer" + i));
    }
}

File: springboot-webflux-demo-Kotlin/src/main/kotlin/com/myproject/webflux/dao/CustomerDao.kt
package com.myproject.webflux.dao

import com.myproject.webflux.dto.Customer
import org.springframework.stereotype.Component
import reactor.core.publisher.Flux
import java.io.InputStream
import java.time.Duration
import java.util.function.Function
import java.util.function.IntConsumer
import java.util.function.IntFunction

import java.util.stream.Collectors

import java.util.stream.IntStream


@Component
class CustomerDao() {
    companion object {
        private fun sleepExecution(i: Int) {
            try {
                Thread.sleep(1000)
            } catch (e: InterruptedException) {
                e.printStackTrace()
            }
        }
    }
    fun getCustomers(): List<Any> {
        return IntStream.rangeClosed(1, 10)
            .peek { i: Int -> sleepExecution(i) }
            .peek { i: Int -> println("processing count : $i") }
            .mapToObj<Any> { i: Int -> Customer(i, "customer$i") }
            .collect(Collectors.toList())
    }
    fun getCustomersStream(): Flux<Customer>{
            return Flux.range(1, 10)
                .delayElements(Duration.ofSeconds(1))
                .doOnNext { i: Int -> println("processing count in stream flow : $i") }
                .map(Function { i: Int -> Customer(i, "customer$i") })
        }
    fun getCustomerList(): Flux<Customer>{
            return Flux.range(1, 50)
                .doOnNext { i: Int -> println("processing count in stream flow : $i") }
                .map(Function { i: Int -> Customer(i, "customer$i") })
        }
}